### PR TITLE
limit parallel downloads for auto repair

### PR DESCRIPTION
**Pull request by ChrisSchinnerl**
This PR should fix the high memory usage when repairing files that are not present on disk.

### CODE REVIEW

File: modules/renter/repair.go [hunk 1]
```diff
@@ -290,14 +288,7 @@ func (r *Renter) managedRepairIteration(rs *repairState) {
...
 		}
 
-		// Don't update gaps for downloaded chunks twice
 		numGaps := chunkStatus.numGaps(rs)
-		if !downloading {
-			// Update the number of gaps for this chunk.
-			rs.gapCounts[chunkStatus.recordedGaps]--
-			rs.gapCounts[numGaps]++
-			chunkStatus.recordedGaps = numGaps
-		}
```
Comment (DavidVorick) [inline modules/renter/repair.go]: pretty sure that this code is critical code, I don't think you want it removed.

Comment (DavidVorick) [reply modules/renter/repair.go]: good rule of thumb for unstable opaque code is to assume that everything is important in delicate an unknown ways. I don't really remember why I put this here,  but I do remember that the gap count stuff was both tricky and also critical

File: modules/renter/repair.go [hunk 2]
```diff
@@ -352,6 +343,11 @@ func (r *Renter) managedRepairIteration(rs *repairState) {
 // managedDownloadChunkData downloads the requested chunk from Sia, for use in
 // the repair loop.
 func (r *Renter) managedDownloadChunkData(rs *repairState, file *file, offset uint64, chunkIndex uint64, chunkID chunkID) ([]byte, error) {
+	// Don't initiate too many downloads to avoid using up all memory
+	if len(rs.downloadingChunks) >= maxScheduledDownloads {
+		return nil, nil
+	}
```
Comment (DavidVorick) [inline modules/renter/repair.go]: I don't think we have any tests that try to repair 5 chunks at the same time, so though the test suite is passing, this construction is not being verified at all.

File: modules/renter/repair.go [hunk 3]
```diff
@@ -290,14 +292,11 @@ func (r *Renter) managedRepairIteration(rs *repairState) {
...
-		// Don't update gaps for downloaded chunks twice
+		// Update the number of gaps for this chunk.
 		numGaps := chunkStatus.numGaps(rs)
-		if !downloading {
-			// Update the number of gaps for this chunk.
-			rs.gapCounts[chunkStatus.recordedGaps]--
-			rs.gapCounts[numGaps]++
-			chunkStatus.recordedGaps = numGaps
-		}
+		rs.gapCounts[chunkStatus.recordedGaps]--
```
Comment (avahowell) [inline modules/renter/repair.go]: what's the rationale for moving this outside of the `if !downloading` branch?

File: api/renterhost_test.go [hunk 4]
```diff
@@ -2331,3 +2331,192 @@ func TestRepairLoopBlocking(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+// TestRemoteFileRepairMassive is similar to TestRemoteFileRepair but uploads
+// more files to find potential deadlocks or crashes
+func TestRemoteFileRepairMassive(t *testing.T) {
```
Comment (avahowell) [inline api/renterhost_test.go]: 👍

File: api/renterhost_test.go [hunk 5]
```diff
@@ -2331,3 +2331,192 @@ func TestRepairLoopBlocking(t *testing.T) {
...
+	}
+
+	// add a few new blocks in order to cause the renter to form contracts with the new host
+	for i := 0; i < 10; i++ {
+		b, err := stNewHost.miner.AddBlock()
+		if err != nil {
+			t.Fatal(err)
+		}
+		for _, tester := range testGroup {
+			err = waitForBlock(b.ID(), tester)
```
Comment (DavidVorick) [inline api/renterhost_test.go]: can just run `synchronizationCheck` here I'm pretty sure it's equivalent to this loop

Comment (ChrisSchinnerl) [reply api/renterhost_test.go]: Are you sure? I copied that code from the other testcase and this loop is also in there. `synchronizationCheck` is already alled a few lines above. (line 2461)

**Pull request merged**